File: src/test/java/im/tox/tox4j/core/NetworkTest.java
package im.tox.tox4j.core;

import im.tox.tox4j.ConnectedListener;
import im.tox.tox4j.ToxCoreImplTestBase;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class NetworkTest extends ToxCoreImplTestBase {

    private static final Logger logger = LoggerFactory.getLogger(NetworkTest.class);

    private static final int TOX_COUNT = 10;

    private void testBootstrap(boolean ipv6Enabled, boolean udpEnabled, String ip, int port, byte[] dhtId) throws Exception {
        String action = String.format("bootstrap to remote node on %s:%d with %s%d",
            ip, port, udpEnabled ? "UDP" : "TCP", ipv6Enabled ? 6 : 4);
        try (ToxCore tox = newTox(ipv6Enabled, udpEnabled)) {
            logger.info("Attempting to {}", action);
            long start = System.currentTimeMillis();
            if (udpEnabled) {
                tox.bootstrap(ip, port, dhtId);
            } else {
                tox.addTcpRelay(ip, port, dhtId);
            }
            ConnectedListener status = new ConnectedListener();
            tox.callbackConnectionStatus(status);
            while (!status.isConnected()) {
                tox.iteration();
                try {
                    Thread.sleep(tox.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            logger.info("Success: {} took {} ms", action, end - start);
        }
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap4_UDP() throws Exception {
        assumeIPv4();
        testBootstrap(false, true, node().ipv4, node().udpPort, node().dhtId);
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap6_UDP() throws Exception {
        assumeIPv6();
        testBootstrap(true, true, node().ipv6, node().udpPort, node().dhtId);
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap4_TCP() throws Exception {
        assumeIPv4();
        testBootstrap(false, false, node().ipv4, node().tcpPort, node().dhtId);
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap6_TCP() throws Exception {
        assumeIPv6();
        testBootstrap(true, false, node().ipv6, node().tcpPort, node().dhtId);
    }

    @Test
    public void testBootstrapSelf() throws Exception {
        // TODO: don't know how to test this on localhost
    }

    @Test(timeout = TIMEOUT)
    public void testLANDiscoveryAll() throws Exception {
        try (ToxList toxes = new ToxList(this, TOX_COUNT)) {
            String action = String.format(
                "Connecting all of %d toxes with LAN discovery", toxes.size());
            logger.info(action);
            long start = System.currentTimeMillis();
            // TODO: Generous timeout required for this; should be made more reliable.
            while (!toxes.isAllConnected()) {
                toxes.iteration();
                try {
                    Thread.sleep(toxes.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            logger.info("{} took {} ms", action, end - start);
        }
    }

    @Test(timeout = TIMEOUT)
    public void testLANDiscoveryAny() throws Exception {
        try (ToxList toxes = new ToxList(this, TOX_COUNT)) {
            String action = String.format(
                "Connecting one of %d toxes with LAN discovery", toxes.size());
            logger.info(action);
            long start = System.currentTimeMillis();
            while (!toxes.isAnyConnected()) {
                toxes.iteration();
                try {
                    Thread.sleep(toxes.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            logger.info("{} took {} ms", action, end - start);
        }
    }

}

File: src/test/java/im/tox/tox4j/core/NetworkTest.scala
package im.tox.tox4j.core

import im.tox.tox4j.DhtNodeSelector.node
import im.tox.tox4j.TestConstants.TIMEOUT
import im.tox.tox4j.ToxCoreTestBase.{ assumeIPv4, assumeIPv6 }
import im.tox.tox4j._
import im.tox.tox4j.core.NetworkTest.logger
import im.tox.tox4j.core.ToxCoreFactory.{ withTox, withToxes }
import org.scalatest.FlatSpec
import org.scalatest.concurrent.Timeouts
import org.scalatest.time.SpanSugar._
import org.slf4j.{ Logger, LoggerFactory }

object NetworkTest {
  private val logger: Logger = LoggerFactory.getLogger(classOf[NetworkTest])
  private val TOX_COUNT: Int = 10
}

final class NetworkTest extends FlatSpec with Timeouts {

  private def testBootstrap(ipv6Enabled: Boolean, udpEnabled: Boolean, ip: String, port: Int, dhtId: Array[Byte]) {
    val action = s"bootstrap to remote node on $ip:$port with ${if (udpEnabled) "UDP" else "TCP"}${if (ipv6Enabled) 6 else 4}"

    withTox(ipv6Enabled, udpEnabled) { tox =>
      logger.info(s"Attempting to $action")
      val start = System.currentTimeMillis

      if (udpEnabled) {
        tox.bootstrap(ip, port, dhtId)
      } else {
        tox.addTcpRelay(ip, port, dhtId)
      }

      val status = new ConnectedListener
      tox.callbackConnectionStatus(status)
      while (!status.isConnected) {
        tox.iteration()
        Thread.sleep(tox.iterationInterval)
      }

      val end = System.currentTimeMillis

      logger.info(s"Success: $action took ${end - start} ms")
    }
  }

  "bootstrap" should "connect with UDP4" in {
    failAfter(TIMEOUT millis) {
      assumeIPv4()
      testBootstrap(ipv6Enabled = false, udpEnabled = true, node.ipv4, node.udpPort, node.dhtId)
    }
  }

  it should "connect with UDP6" in {
    failAfter(TIMEOUT millis) {
      assumeIPv6()
      testBootstrap(ipv6Enabled = true, udpEnabled = true, node.ipv6, node.udpPort, node.dhtId)
    }
  }

  it should "connect with TCP4" in {
    failAfter(TIMEOUT millis) {
      assumeIPv4()
      testBootstrap(ipv6Enabled = false, udpEnabled = false, node.ipv4, node.tcpPort, node.dhtId)
    }
  }

  it should "connect with TCP6" in {
    failAfter(TIMEOUT millis) {
      assumeIPv6()
      testBootstrap(ipv6Enabled = true, udpEnabled = false, node.ipv6, node.tcpPort, node.dhtId)
    }
  }

  "LAN discovery" should "connect all nodes" in {
    failAfter(TIMEOUT millis) {
      withToxes(NetworkTest.TOX_COUNT) { toxes =>
        val action = s"Connecting all of ${toxes.size} toxes with LAN discovery"
        logger.info(action)

        val start = System.currentTimeMillis

        while (!toxes.isAllConnected) {
          toxes.iteration()
          Thread.sleep(toxes.iterationInterval)
        }

        val end = System.currentTimeMillis

        logger.info(s"$action took ${end - start} ms")
      }
    }
  }

  it should "connect at least one instance" in {
    failAfter(TIMEOUT millis) {
      withToxes(NetworkTest.TOX_COUNT) { toxes =>
        val action = s"Connecting one of ${toxes.size} toxes with LAN discovery"
        logger.info(action)

        val start = System.currentTimeMillis

        while (!toxes.isAnyConnected) {
          toxes.iteration()
          try {
            Thread.sleep(toxes.iterationInterval)
          } catch {
            case e: InterruptedException =>
          }
        }

        val end = System.currentTimeMillis

        logger.info(s"$action took ${end - start} ms")
      }
    }
  }

}
